interfaces/hotplug/udevadm.go

```diff
@@ -0,0 +1,110 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2018 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package hotplug
+
+import (
+	"bufio"
+	"fmt"
+	"os/exec"
+	"regexp"
+	"strings"
+)
+
+var udevadmBin = `/sbin/udevadm`
+var udevadmRe = regexp.MustCompile("^([A-Z]): (.*)$")
+
+// RunUdevadm enumerates all devices by parsing 'udevadm info -e' command output and reports them
+// via devices channel. The devices channel gets closed to indicate that all devices were processed.
+// Parsing errors are reported to parseErrors channel. Fatal errors encountered when starting udevadm
+// are reported by the error return value.
+func RunUdevadm(devices chan<- *HotplugDeviceInfo, parseErrors chan<- error) error {
+	cmd := exec.Command(udevadmBin, "info", "-e")
+	stdout, err := cmd.StdoutPipe()
+	if err != nil {
+		close(devices)
+		return err
+	}
+
+	rd := bufio.NewScanner(stdout)
+	if err := cmd.Start(); err != nil {
+		close(devices)
+		return err
+	}
+
+	go func() {
+		defer close(devices)
+
+		env := make(map[string]string)
+		for rd.Scan() {
+			line := rd.Text()
+			if line == "" {
+				if len(env) > 0 {
+					outputDevice(env, devices, parseErrors)
+					env = make(map[string]string)
+				}
+				continue
+			}
+			match := udevadmRe.FindStringSubmatch(line)
```

Comment:
Don't think you need to use a regex here. If we're really only interested in parsing property lines, something like this will suffice:
```
if strings.HasPrefix("E: ") {
    if kv := strings.SplitN(line[3:], "=", 2); kv == 2 {
        env[kv[0]] = kv[1]
    } else {
        parseErrors <- fmt.Errorf("failed to parse udevadm output %q", line)        
    }
}
```

Reply:
Fair enough, got rid of regex.